data/transactions/logic/eval.go

```diff
@@ -4473,10 +4556,65 @@ func asaReference(cx *EvalContext, ref uint64, foreign bool) (basics.AssetIndex,
 			return basics.AssetIndex(ref), nil
 		}
 	}
-	return basics.AssetIndex(0), fmt.Errorf("invalid Asset reference %d", ref)
+	return 0, fmt.Errorf("invalid Asset reference %d", ref)
 
 }
 
+func (cx *EvalContext) holdingReference(account stackValue, ref uint64) (basics.Address, basics.AssetIndex, error) {
+	if cx.version >= resourceSharingVersion {
+		var addr basics.Address
+		var err error
+		if account.Bytes != nil {
+			addr, err = account.address()
+		} else {
+			addr, err = cx.txn.Txn.AddressByIndex(account.Uint, cx.txn.Txn.Sender)
+		}
+		if err != nil {
+			return basics.Address{}, 0, err
+		}
+		aid := basics.AssetIndex(ref)
```

Comment:
Do you mean that it's important that we're checking the address first? I don't think it should matter, the only way `err != nil` at line 4572 is because the thing passed by the user is fundamentally not an address. Either it's a `[]byte` but not 32 bytes long, or it's an integer above len(txn.Accounts).  This isn't an availability check. Only on line 4576 are we doing the "real" check, and that check is checking the account and asset "simultaneously". That is, it's looking in the `cx.available` structure.

Reply:
I meant if it looks first into ForeignAssets, then aidx=1 might not be resolved right (it could be an index pointing to aidx=X)

Reply:
I see. Yes, we decided a while back that the first interpretation is as an asset id. If that's available, then that is the interpretation. If it's not available, it is an index in ForeignAssets.

In the real chains, this can never matter, since the low assets / apps don't exist. But this way allows txns to be unambiguous by specifying the asset id.  I'm not sure we have a test though.  I'll check and (maybe) add.

Reply:
See `TestAppDisambiguation` and `TestAssetDisambiguation` in `evalStateful_test.go`.

Thinking about this issue is leading us to remove the dual meanings in v9 and above, but these tests are good to have for earlier version. Thanks.